### ISSUE TITLE
ConcurrentModificationException

### ISSUE THREAD

**Issue by hurzl** [SecUSo/privacy-friendly-weather#148]
when switching tab/location

```
01-16 01:42:31.768 30812 30812 E AndroidRuntime: FATAL EXCEPTION: main
01-16 01:42:31.768 30812 30812 E AndroidRuntime: Process: org.secuso.privacyfriendlyweather, PID: 30812
01-16 01:42:31.768 30812 30812 E AndroidRuntime: java.util.ConcurrentModificationException
01-16 01:42:31.768 30812 30812 E AndroidRuntime: 	at java.util.ArrayList$Itr.next(ArrayList.java:860)
01-16 01:42:31.768 30812 30812 E AndroidRuntime: 	at org.secuso.privacyfriendlyweather.ui.updater.ViewUpdater.updateCurrentWeatherData(ViewUpdater.java:28)
01-16 01:42:31.768 30812 30812 E AndroidRuntime: 	at org.secuso.privacyfriendlyweather.weather_api.open_weather_map.ProcessOwmForecastOneCallAPIRequest.processSuccessScenario(ProcessOwmForecastOneCallAPIRequest.java:124)
01-16 01:42:31.768 30812 30812 E AndroidRuntime: 	at org.secuso.privacyfriendlyweather.http.VolleyHttpRequest$1.onResponse(VolleyHttpRequest.java:82)
01-16 01:42:31.768 30812 30812 E AndroidRuntime: 	at org.secuso.privacyfriendlyweather.http.VolleyHttpRequest$1.onResponse(VolleyHttpRequest.java:79)
01-16 01:42:31.768 30812 30812 E AndroidRuntime: 	at com.android.volley.toolbox.StringRequest.deliverResponse(StringRequest.java:82)
01-16 01:42:31.768 30812 30812 E AndroidRuntime: 	at com.android.volley.toolbox.StringRequest.deliverResponse(StringRequest.java:29)
01-16 01:42:31.768 30812 30812 E AndroidRuntime: 	at com.android.volley.ExecutorDelivery$ResponseDeliveryRunnable.run(ExecutorDelivery.java:102)
01-16 01:42:31.768 30812 30812 E AndroidRuntime: 	at android.os.Handler.handleCallback(Handler.java:883)
01-16 01:42:31.768 30812 30812 E AndroidRuntime: 	at android.os.Handler.dispatchMessage(Handler.java:100)
01-16 01:42:31.768 30812 30812 E AndroidRuntime: 	at android.os.Looper.loop(Looper.java:214)
01-16 01:42:31.768 30812 30812 E AndroidRuntime: 	at android.app.ActivityThread.main(ActivityThread.java:7356)
01-16 01:42:31.768 30812 30812 E AndroidRuntime: 	at java.lang.reflect.Method.invoke(Native Method)
01-16 01:42:31.768 30812 30812 E AndroidRuntime: 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:491)
01-16 01:42:31.768 30812 30812 E AndroidRuntime: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:940)
```

**Comment by sleep-yearning** [comment]
Thanks for reporting this.
Does it happen consistently or just at some times?

**Comment by hurzl** [comment]
almost every time when weather is updated

**Comment by sleep-yearning** [comment]
Note for resolving: Probably a change in the ViewUpdater subscriber list, while it is iterated for weather updating.
This would mean that the ForecastActivity is resumed/paused or a WeatherCityFragment is attached/detached while the collected data is propagated.

**Comment by hurzl** [comment]
Two versions later same crash, same linenumbers ...

**Comment by GenosseFlosse** [comment]
I encounter the same issue when swiping from right to left in order to change the current city.
```
java.util.ConcurrentModificationException
	at java.util.ArrayList$Itr.next(ArrayList.java:860)
	at org.secuso.privacyfriendlyweather.ui.updater.ViewUpdater.updateCurrentWeatherData(ViewUpdater.java:28)
	at org.secuso.privacyfriendlyweather.weather_api.open_weather_map.ProcessOwmForecastOneCallAPIRequest.processSuccessScenario(ProcessOwmForecastOneCallAPIRequest.java:123)
	at org.secuso.privacyfriendlyweather.http.VolleyHttpRequest$1.onResponse(VolleyHttpRequest.java:82)
	at org.secuso.privacyfriendlyweather.http.VolleyHttpRequest$1.onResponse(VolleyHttpRequest.java:79)
	at com.android.volley.toolbox.StringRequest.deliverResponse(StringRequest.java:82)
	at com.android.volley.toolbox.StringRequest.deliverResponse(StringRequest.java:29)
	at com.android.volley.ExecutorDelivery$ResponseDeliveryRunnable.run(ExecutorDelivery.java:102)
	at android.os.Handler.handleCallback(Handler.java:883)
	at android.os.Handler.dispatchMessage(Handler.java:100)
	at android.os.Looper.loop(Looper.java:224)
	at android.app.ActivityThread.main(ActivityThread.java:7562)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:539)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:950)
```

**Comment by woheller69** [comment]
I think, I fixed this with

**if (subscribers.contains(sub))** {        //Check if list element has not been removed in the meantime; Bugfix for Concurrent Modification Exception
                    sub.processNewCurrentWeatherData(data);
                }

https://github.com/woheller69/weather/commit/7a89e3055095bcf3ae8b65ffde92cf4da3a354e7

Version 1.9 of my RadarWeather including this fix should be published by F-Droid soon.
Maybe @hurzl can test it there. I only rarely got this problem

**Comment by woheller69** [comment]
V1.9 is now available on F-Droid.

https://f-droid.org/de/packages/org.woheller69.weather/

Maybe you can try if you still get the error there. If not this fix should also be integrated here.

**Comment by hurzl** [comment]
In ViewUpdater, use:

```
    public static void updateCurrentWeatherData(CurrentWeatherData data) {
        ArrayList<IUpdateableCityUI> subcopy = new ArrayList<>(subscribers);
        for (IUpdateableCityUI sub : subcopy) {
            sub.processNewWeatherData(data);
        }
    }
```
etc.

**Comment by sleep-yearning** [comment]
Tanks for the suggestions to both of you. I am hesitant to implement them because I don't know if it will fix it. The issue was hard to replicate so far. Can you confirm @hurzl that the issues go away with that change? And that the relevant views still get updated?
I'm also unsure about the other change because it seems that the List itself is complaining and not the missing object is throwing the error.

**Comment by hurzl** [comment]
it fixes the problem in @woheller69's version

**Comment by hurzl** [comment]
What's the problem?

**Comment by woheller69** [comment]
My RadarWeather V2.0 including this fix and a new widget is now ready for download on F-Droid.
New feature next week: a click on a weekday will scroll course of day to this position

**Comment by sleep-yearning** [comment]
should be solved now with c689923e32fc28575f43ad630af4dff328649f55
I implemented the changes from @hurzl but not @woheller69, since no error occurred without it.
